File: app/src/main/java/il/ronmad/speedruntimer/CountdownPreference.java
package il.ronmad.speedruntimer;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;

public class CountdownPreference extends DialogPreference {

    private long countdown;
    private EditText hoursInput;
    private EditText minutesInput;
    private EditText secondsInput;
    private EditText millisInput;

    public CountdownPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public CountdownPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        countdown = 0L;
    }

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);
        View view = LayoutInflater.from(getContext()).inflate(R.layout.edit_time_dialog, null);
        hoursInput = view.findViewById(R.id.hours);
        minutesInput = view.findViewById(R.id.minutes);
        secondsInput = view.findViewById(R.id.seconds);
        millisInput = view.findViewById(R.id.milliseconds);
        Util.setEditTextsFromTime(countdown,
                hoursInput, minutesInput, secondsInput, millisInput);
        builder.setView(view)
                .setPositiveButton(R.string.save, (dialogInterface, i) ->  {
                    countdown = Util.getTimeFromEditTexts(hoursInput, minutesInput, secondsInput, millisInput);
                    persistLong(countdown);
                    setSummary("Timer starts at " + Util.getFormattedTime(-1 * countdown));
                })
                .setNegativeButton(R.string.pb_clear, this)
                .setNeutralButton(android.R.string.cancel, this);
    }

    @Override
    protected void showDialog(Bundle state) {
        super.showDialog(state);
        AlertDialog dialog = (AlertDialog) getDialog();
        dialog.getButton(DialogInterface.BUTTON_NEGATIVE).setOnClickListener(view -> {
            hoursInput.setText("");
            minutesInput.setText("");
            secondsInput.setText("");
            millisInput.setText("");
        });
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return (long) a.getInt(index, 0);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        countdown = (restorePersistedValue ?
                getPersistedLong(countdown) : (long) defaultValue);
    }
}

File: app/src/main/kotlin/il/ronmad/speedruntimer/CountdownPreference.kt
package il.ronmad.speedruntimer

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.preference.DialogPreference
import androidx.preference.Preference
import androidx.preference.Preference.OnPreferenceChangeListener
import androidx.preference.Preference.SummaryProvider

class CountdownPreference : DialogPreference {

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context) : super(context)

    var countdown: Long = 0L
        set(value) {
            if (callChangeListener(value)) {
                field = value
                persistLong(value)
                notifyChanged()
            }
        }

    init {
        onPreferenceChangeListener = OnPreferenceChangeListener { _: Preference, value: Any ->
            (value as Long) != countdown
        }
        summaryProvider = SummaryProvider { _: DialogPreference ->
            "Timer starts at ${(-countdown).getFormattedTime()}"
        }
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int) = a.getInt(index, 0).toLong()

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        countdown = if (restorePersistedValue)
            getPersistedLong(countdown) else ((defaultValue as? Long) ?: 0)
    }
}
